File: examples/javascript/social/get-friends-list.js
/**
 * Get Friends List
 *
 * Retrieves and displays your complete friends list.
 *
 * Usage in browser console:
 * 1. Navigate to geoguessr.com (must be logged in)
 * 2. Paste this code into the console
 * 3. Call: getFriendsList()
 */

async function getFriendsList() {
    try {
        const response = await fetch('https://www.geoguessr.com/api/v3/social/friends', {
            credentials: 'include'
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const friends = await response.json();

        if (friends.length === 0) {
            console.log('You have no friends yet. Add some friends to see them here!');
            return [];
        }

        // Display friends list
        console.log(`\n👥 Your Friends (${friends.length} total)\n`);

        const tableData = friends.map(friend => ({
            Username: friend.nick,
            Level: friend.progress?.level || 'N/A',
            Country: friend.countryCode?.toUpperCase() || 'N/A',
            Pro: friend.isProUser ? '✅' : '❌',
            Online: friend.isOnline ? '🟢' : '⚪',
            ID: friend.userId
        }));

        console.table(tableData);

        // Statistics
        const onlineFriends = friends.filter(f => f.isOnline).length;
        const proFriends = friends.filter(f => f.isProUser).length;
        const avgLevel = friends
            .filter(f => f.progress?.level)
            .reduce((sum, f) => sum + f.progress.level, 0) / friends.length;

        console.log(`\n📊 Statistics:`);
        console.log(`Total Friends: ${friends.length}`);
        console.log(`Online Now: ${onlineFriends}`);
        console.log(`Pro Users: ${proFriends} (${((proFriends / friends.length) * 100).toFixed(1)}%)`);
        console.log(`Average Level: ${avgLevel.toFixed(1)}`);

        // Group by country
        const countries = {};
        friends.forEach(f => {
            if (f.countryCode) {
                countries[f.countryCode] = (countries[f.countryCode] || 0) + 1;
            }
        });

        console.log(`\n🌍 Countries Represented:`);
        Object.entries(countries)
            .sort((a, b) => b[1] - a[1])
            .slice(0, 5)
            .forEach(([code, count]) => {
                console.log(`${code.toUpperCase()}: ${count}`);
            });

        return friends;

    } catch (error) {
        console.error('❌ Error fetching friends:', error.message);
        return null;
    }
}

// Example usage:
// getFriendsList();

// Find a specific friend:
// const friends = await getFriendsList();
// const friend = friends.find(f => f.nick === 'username');

File: examples/python/social/get_friends_list.py
"""
Get Friends List

Retrieves and displays your complete friends list.

Usage:
    python get_friends_list.py

Requirements:
    pip install requests
"""

import requests
import os
from collections import Counter


def get_friends_list(cookie):
    """
    Fetch and display your friends list.

    Args:
        cookie (str): Your _ncfa cookie value

    Returns:
        list: List of friends
    """
    try:
        response = requests.get(
            'https://www.geoguessr.com/api/v3/social/friends',
            cookies={'_ncfa': cookie}
        )
        response.raise_for_status()

        friends = response.json()

        if len(friends) == 0:
            print('You have no friends yet. Add some friends to see them here!')
            return []

        # Display friends list
        print(f"\n👥 Your Friends ({len(friends)} total)\n")
        print(f"{'Username':<20} {'Level':<8} {'Country':<10} {'Pro':<5} {'Online'}")
        print('-' * 60)

        for friend in friends:
            username = friend['nick'][:19]
            level = friend.get('progress', {}).get('level', 'N/A')
            country = friend.get('countryCode', 'N/A').upper()
            pro = '✅' if friend.get('isProUser') else '❌'
            online = '🟢' if friend.get('isOnline') else '⚪'

            print(f"{username:<20} {level:<8} {country:<10} {pro:<5} {online}")

        # Statistics
        online_count = sum(1 for f in friends if f.get('isOnline'))
        pro_count = sum(1 for f in friends if f.get('isProUser'))

        levels = [f.get('progress', {}).get('level', 0) for f in friends if f.get('progress', {}).get('level')]
        avg_level = sum(levels) / len(levels) if levels else 0

        print(f"\n📊 Statistics:")
        print(f"Total Friends: {len(friends)}")
        print(f"Online Now: {online_count}")
        print(f"Pro Users: {pro_count} ({(pro_count / len(friends) * 100):.1f}%)")
        print(f"Average Level: {avg_level:.1f}")

        # Group by country
        countries = [f.get('countryCode') for f in friends if f.get('countryCode')]
        country_counts = Counter(countries).most_common(5)

        print(f"\n🌍 Top Countries:")
        for country, count in country_counts:
            print(f"{country.upper()}: {count}")

        return friends

    except requests.exceptions.HTTPError as e:
        print(f"❌ HTTP Error: {e}")
        return None
    except requests.exceptions.RequestException as e:
        print(f"❌ Request Error: {e}")
        return None


def main():
    # Get credentials from environment variables
    cookie = os.getenv('GEOGUESSR_COOKIE')
    if not cookie:
        print("❌ Error: GEOGUESSR_COOKIE environment variable not set")
        print("Set it with: export GEOGUESSR_COOKIE='your_cookie_value'")
        return

    # Fetch friends
    friends = get_friends_list(cookie)

    if friends:
        print(f"\n✅ Retrieved {len(friends)} friends!")


if __name__ == '__main__':
    main()
